File: python/1590.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n = len(nums)
        total = sum(nums)
        if total % p == 0:
            return 0
        target = total % p
        prefix = 0
        last_index = {0: -1} # dictionary
        min_len = n
        for i in range(n):
            prefix = (prefix + nums[i]) % p
            need = (prefix - target + p) % p
            if need in last_index:
                min_len = min(min_len, i - last_index[need])
            last_index[prefix] = i
        return min_len if min_len != n else -1

File: cpp/1590.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        int n = nums.size();
        
        // Calculate total sum modulo p
        long long total = accumulate(nums.begin(), nums.end(), 0LL);
        
        // If already divisible, return 0
        int target = total % p;
        if(target == 0) {
            return 0;
        }
        
        // Use prefix sum with modulo to find smallest subarray with remainder = target
        unordered_map<int, int> lastIndex; // maps prefix_sum % p -> last index
        lastIndex[0] = -1; // prefix sum 0 at index -1 (before start)
        
        long long prefix = 0;
        int minLen = n;
        
        for(int i = 0; i < n; i++) {
            prefix = (prefix + nums[i]) % p;
            
            // We want to find a subarray with remainder = target
            // If prefix % p = r, we need previous prefix % p = (r - target + p) % p
            int need = (prefix - target + p) % p;
            
            if(lastIndex.count(need)) {
                minLen = min(minLen, i - lastIndex[need]);
            }
            
            lastIndex[prefix] = i;
        }
        
        return minLen == n ? -1 : minLen;
    }
};